src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RegisterShopUserHandlerSpec.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace spec\Sylius\Bundle\ApiBundle\CommandHandler;
+
+use Doctrine\Persistence\ObjectManager;
+use PhpSpec\ObjectBehavior;
+use Sylius\Bundle\ApiBundle\Command\RegisterShopUser;
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\Core\Repository\CustomerRepositoryInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\User\Canonicalizer\CanonicalizerInterface;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+final class RegisterShopUserHandlerSpec extends ObjectBehavior
+{
+    function let(
+        CanonicalizerInterface $canonicalizer,
+        FactoryInterface $shopUserFactory,
+        FactoryInterface $customerFactory,
+        CustomerRepositoryInterface $customerRepository,
+        ObjectManager $shopUserManager
+    ): void {
+        $this->beConstructedWith(
+            $canonicalizer,
+            $shopUserFactory,
+            $customerFactory,
+            $customerRepository,
+            $shopUserManager
+        );
+    }
+
+    function it_is_a_message_handler(): void
+    {
+        $this->shouldImplement(MessageHandlerInterface::class);
+    }
+
+    function it_creates_a_customer_and_user_with_given_data(
+        CanonicalizerInterface $canonicalizer,
+        FactoryInterface $shopUserFactory,
+        FactoryInterface $customerFactory,
+        CustomerRepositoryInterface $customerRepository,
+        ObjectManager $shopUserManager,
+        ShopUserInterface $shopUser,
+        CustomerInterface $customer
+    ): void {
+        $canonicalizer->canonicalize('WILL.SMITH@example.com')->willReturn('will.smith@example.com');
+        $customerRepository->findOneBy(['emailCanonical' => 'will.smith@example.com'])->willReturn(null);
+
+        $shopUserFactory->createNew()->willReturn($shopUser);
+        $customerFactory->createNew()->willReturn($customer);
+
+        $shopUser->setPlainPassword('iamrobot')->shouldBeCalled();
+
+        $customer->setFirstName('Will')->shouldBeCalled();
+        $customer->setLastName('Smith')->shouldBeCalled();
+        $customer->setEmail('WILL.SMITH@example.com')->shouldBeCalled();
+        $customer->setPhoneNumber('+13104322400')->shouldBeCalled();
+        $customer->setUser($shopUser)->shouldBeCalled();
+
+        $shopUserManager->persist($shopUser)->shouldBeCalled();
+
+        $this->__invoke(new RegisterShopUser('Will', 'Smith', 'WILL.SMITH@example.com', 'iamrobot', '+13104322400'));
```

Comment:
Just as a proposal for different style (I prefer this one in invokables), but feel free to reject it
```suggestion
        $this(new RegisterShopUser('Will', 'Smith', 'WILL.SMITH@example.com', 'iamrobot', '+13104322400'));
```